core/kotlinx-coroutines-test/src/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,294 @@
+     * or [advanceUntilIdle].
+     */
+    @ExperimentalCoroutinesApi
+    fun resumeDispatcher()
+}
+
+/**
+ * Thrown when a test has completed and there are tasks that are not completed or cancelled.
+ */
+@ExperimentalCoroutinesApi
+class UncompletedCoroutinesError(message: String, cause: Throwable? = null): AssertionError(message, cause)
+
+/**
+ * [CoroutineDispatcher] that can be used in tests for both immediate and lazy execution of coroutines.
+ *
+ * By default, [TestCoroutineDispatcher] will be immediate. That means any tasks scheduled to be run without delay will
+ * be immediately executed. If they were scheduled with a delay, the virtual clock-time must be advanced via one of the
+ * methods on [DelayController]
+ *
+ * When swiched to lazy execution any coroutines started via [launch] or [async] will
+ * not execute until a call to [DelayController.runCurrent] or the virtual clock-time has been advanced via one of the
+ * methods on [DelayController].
+ *
+ * @see DelayController
+ */
+@ExperimentalCoroutinesApi
+class TestCoroutineDispatcher:
+        CoroutineDispatcher(),
+        Delay,
+        DelayController {
+
+    private var dispatchImmediately = true
+        set(value) {
+            field = value
+            if (value) {
+                // there may already be tasks from setup code we need to run
+                advanceUntilIdle()
+            }
+        }
+
+    // The ordered queue for the runnable tasks.
+    private val queue = ThreadSafeHeap<TimedRunnable>()
+
+    // The per-scheduler global order counter.
+    private var counter = AtomicLong(0)
+
+    // Storing time in nanoseconds internally.
+    private var time = AtomicLong(0)
+
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        if (dispatchImmediately) {
+            block.run()
+        } else {
+            post(block)
+        }
+    }
+
+    override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) {
+        postDelayed(CancellableContinuationRunnable(continuation) { resumeUndispatched(Unit) }, timeMillis)
+    }
+
+    override fun invokeOnTimeout(timeMillis: Long, block: Runnable): DisposableHandle {
+        val node = postDelayed(block, timeMillis)
+        return object : DisposableHandle {
+            override fun dispose() {
+                queue.remove(node)
+            }
+        }
+    }
+
+    override fun toString(): String {
+        return "TestCoroutineDispatcher[currentTime=${time}ms, queued=${queue.size}]"
+    }
+
+    private fun post(block: Runnable) =
+            queue.addLast(TimedRunnable(block, counter.getAndIncrement()))
+
+    private fun postDelayed(block: Runnable, delayTime: Long) =
+            TimedRunnable(block, counter.getAndIncrement(), time.get() + delayTime)
+                    .also {
+                        queue.addLast(it)
+                    }
+
+
+    private fun doActionsUntil(targetTime: Long) {
```

Comment:
can we make this return boolean (or int, # of executed coroutines blocks) OR provide a function to check if idle?

I'm using the old one in another test where i have multiple TestCoroutineContexts. I want to have a function that progresses all of them (and they enqueue messages into eachother) and i cannot write a custom function without reflecting into the queue.

Reply:
to be clear, the version I need is expected to ignore delayed tasks if time is not ready.
Maybe i'm  missing a function that already exists :/.
Not proud of it but here is what i'm doing there:
```
private fun triggerAllActions() {
        do {
            mainContext.triggerActions()
            testContext.triggerActions()
            val allIdle = listOf(mainContext, testContext).all {
                it.isIdle()
            }
        } while (!allIdle)
    }

    private fun TestCoroutineContext.isIdle() : Boolean {
        val queueField = this::class.java
            .getDeclaredField("queue")
        queueField.isAccessible = true
        val queue = queueField.get(this)
        val peekMethod = queue::class.java
            .getDeclaredMethod("peek")
        val nextTask = peekMethod.invoke(queue) ?: return true
        val timeField = nextTask::class.java.getDeclaredField("time")
        timeField.isAccessible = true
        val time = timeField.getLong(nextTask)
        return time > now()
    }
```